Fix App Search crawler URL.

afoucret opened the pull request:
## Summary

When clicking on the link to configure the webcrawler we are redirected to the following URL : `http://localhost:3002/as/engines/foo/crawler` in the legacy standalone UI. This result in a routing error in App Search:

<img width="1660" alt="Capture d’écran 2021-06-28 à 15 05 48" src="https://user-images.githubusercontent.com/529238/123641557-aa257000-d822-11eb-84fe-0034f88c3ea9.png">

<img width="767" alt="Capture d’écran 2021-06-28 à 15 06 15" src="https://user-images.githubusercontent.com/529238/123641682-caedc580-d822-11eb-80dc-93395acc8c00.png">

This is because of a missing anchor in the generated URL that should be `http://localhost:3002/as#/engines/foo/crawler`

This PR is intended to fix this bug,

### Checklist
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134568/)
* Commit: e7b014a4e4abac97f895406254db838bf52054b9
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-103479/e7b014a4e4abac97f895406254db838bf52054b9)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134568/execution/node/190/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134568/","number":"134568","commit":"e7b014a4e4abac97f895406254db838bf52054b9"}]}
PIPELINE-->

JasonStoltz commented on x-pack/plugins/enterprise_search/public/applications/shared/enterprise_search_url/external_url.test.ts:
```diff
@@ -28,16 +28,22 @@ describe('Enterprise Search external URL helpers', () => {
       expect(getEnterpriseSearchUrl()).toEqual('http://localhost:3002');
       expect(getEnterpriseSearchUrl('/login')).toEqual('http://localhost:3002/login');
+      expect(getEnterpriseSearchUrl('/login', '/a')).toEqual('http://localhost:3002/a/login');
+      expect(getEnterpriseSearchUrl('/login', '/a', true)).toEqual('http://localhost:3002/bar#/a');
```
```suggestion
      expect(getEnterpriseSearchUrl('/login', '/a', true)).toEqual('http://localhost:3002/a#/login');
```

afoucret replied:
This is fixed. Thank you

JasonStoltz commented:
@afoucret Just FYI, I added some tags to this PR, as CI will complain otherwise. PRs always require:
- A target version label
- A release note tag (or release_note:skip, which is what we typically do)

I also added the `auto-backport` so that it will be backported automatically to 7.x and included in 7.14.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134611/)
* Commit: 727a9161d0c9291f39a3ed3f10350fecb36b01a8
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Quick Commit Checks](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134611/execution/node/407/log)

### History
* :broken_heart: [Build #134568](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134568/) failed e7b014a4e4abac97f895406254db838bf52054b9

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134568/","number":"134568","commit":"e7b014a4e4abac97f895406254db838bf52054b9"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134611/","number":"134611","commit":"727a9161d0c9291f39a3ed3f10350fecb36b01a8"}]}
PIPELINE-->

JasonStoltz commented on x-pack/plugins/enterprise_search/public/applications/shared/enterprise_search_url/external_url.test.ts:
```diff
@@ -28,16 +28,22 @@ describe('Enterprise Search external URL helpers', () => {
-    it('generates a public App Search URL', () => {
-      expect(getAppSearchUrl()).toEqual('http://localhost:3002/as');
-      expect(getAppSearchUrl('/path')).toEqual('http://localhost:3002/as/path');
+    describe('App Search URL helper', () => {
```
This is not a blocking comment, but just calling it out. Why did you decided to add a `describe` block to the app search helper but not the workplace search helper?

afoucret replied:
I removed it. Initially, I wanted to add a `describe` for the method without anchor and another one for the method with it

JasonStoltz reviewed (approved):
I'm good with these changes. Will QA very quickly once my Kibana actually starts 🕐 .

constancecchen commented on x-pack/plugins/enterprise_search/public/applications/app_search/components/crawler/crawler_landing.tsx:
```diff
@@ -63,7 +63,7 @@ export const CrawlerLanding: React.FC = () => (
           fill
           color="primary"
-          href={getAppSearchUrl(generateEnginePath(ENGINE_CRAWLER_PATH))}
+          href={getAppSearchUrl(generateEnginePath(ENGINE_CRAWLER_PATH), true)}
```
```suggestion
          href={getAppSearchUrl('#' + generateEnginePath(ENGINE_CRAWLER_PATH), true)}
```

constancecchen commented on x-pack/plugins/enterprise_search/public/applications/shared/enterprise_search_url/external_url.ts:
```diff
@@ -27,12 +27,16 @@ export const externalUrl = {
 };
-export const getAppSearchUrl = (path: string = ''): string => {
-  return getEnterpriseSearchUrl('/as' + path);
+export const getAppSearchUrl = (path: string = '', useAnchor: boolean = false): string => {
+  return getEnterpriseSearchUrl(path, '/as', useAnchor);
```
```suggestion
export const getAppSearchUrl = (path: string = ''): string => {
  return getEnterpriseSearchUrl(path, '/as#');
```

TBH, I'm not a huge fan of how our App Search links work and how some of them require `#` anchors and some don't (because they correctly redirect).

So if I had a preferred option between the two, I'd probably choose this one for simplicity, since EntSearch fixed login redirects not preserving hashes a while back.

constancecchen reviewed (changes requested):
I super hate to do this, but since the Crawer Landing is temporary and we will eventually not have any URL links going out to the standalone UI, I'm not a fan of overcomplicating the current set of helpers.

I would propose either (but not both) of the below solutions as simpler approaches that have no helper API updates. Up to you as well whether you'd prefer to revert changes or open a new PR

byronhulcher commented:
Sorry to everyone for missing this

JasonStoltz commented:
@constancecchen

? and we will eventually not have any URL links going out to the standalone UI, I'm not a fan of overcomplicating the current set of helpers.

FWIW, I'd use the same logic to say, it's good enough as it is because it's all going away eventually anyway.

constancecchen commented:
> FWIW, I'd use the same logic to say, it's good enough as it is because it's all going away eventually anyway.

The URL **helpers** are not going away - the Search UI external links have to use them. So these helpers have to remain in perpetuity, and I'd strongly rather not add another API/param to them over simply modifying the URL with an extra static `#`string

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134618/)
* Commit: e9ffce65e1321c4e959e8a9c6757e09d09c48214
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-103479/e9ffce65e1321c4e959e8a9c6757e09d09c48214)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/f89dc9cc31aa139f47d83133bdc10145ab7c0468) | [after](https://github.com/elastic/kibana/commit/e9ffce65e1321c4e959e8a9c6757e09d09c48214) | diff |
| --- | --- | --- | --- |
| `enterpriseSearch` | 2.1MB | 2.1MB | +223.0B |

### History
* :broken_heart: [Build #134568](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134568/) failed e7b014a4e4abac97f895406254db838bf52054b9
* :broken_heart: [Build #134611](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134611/) failed 727a9161d0c9291f39a3ed3f10350fecb36b01a8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134611/","number":"134611","commit":"727a9161d0c9291f39a3ed3f10350fecb36b01a8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134568/","number":"134568","commit":"e7b014a4e4abac97f895406254db838bf52054b9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134618/","number":"134618","commit":"e9ffce65e1321c4e959e8a9c6757e09d09c48214"}]}
PIPELINE-->

afoucret commented:
I rebased the PR to remove the commits of the URL helpers overcomplicated modifications.
Instead, I added an anchor directly into the crawler landing page as suggested by @constancecchen 

I also agree that we should manage better redirect in App Search instead of hacking links in the frontend.

JasonStoltz reviewed (approved):
Much simpler!

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135063/)
* Commit: 06362790adb334aae6de0952b436fea21ad7d4e5
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-103479/06362790adb334aae6de0952b436fea21ad7d4e5)
* [Documentation Changes](https://kibana_103479.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/af62b05e3a72788a88ea97e7699365fffff45fac) | [after](https://github.com/elastic/kibana/commit/06362790adb334aae6de0952b436fea21ad7d4e5) | diff |
| --- | --- | --- | --- |
| `enterpriseSearch` | 2.1MB | 2.1MB | +4.0B |

### History
* :green_heart: [Build #134618](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134618/) succeeded e9ffce65e1321c4e959e8a9c6757e09d09c48214
* :broken_heart: [Build #134568](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134568/) failed e7b014a4e4abac97f895406254db838bf52054b9
* :broken_heart: [Build #134611](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134611/) failed 727a9161d0c9291f39a3ed3f10350fecb36b01a8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134611/","number":"134611","commit":"727a9161d0c9291f39a3ed3f10350fecb36b01a8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134568/","number":"134568","commit":"e7b014a4e4abac97f895406254db838bf52054b9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/134618/","number":"134618","commit":"e9ffce65e1321c4e959e8a9c6757e09d09c48214"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135063/","number":"135063","commit":"06362790adb334aae6de0952b436fea21ad7d4e5"}]}
PIPELINE-->

afoucret commented:
@constancecchen It seems that you need to approve the PR since you required some changes.
Let me know if this is OK.

Also, I have enabled auto-merge on it. So it will be merged automatically if you approve it.

constancecchen reviewed (approved):
❤️ Thanks for being willing to simplify Aurelien, and super appreciate this awesome catch on your part!!

kibanamachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [7.x](https://github.com/elastic/kibana/pull/103698)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/103698">](https://github.com/elastic/kibana/pull/103698) |

This backport PR will be merged automatically after passing CI.

afoucret commented:
@constancecchen Your welcome. 
And it was fun to make my first contrib to Kibana. I feel like if I got an upgrade to my skillset!
